pkg/logs/input/docker/launcher.go

```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
 	}
 
 	// keep the tailer in track to stop it later on
-	l.tailers[containerID] = tailer
+	l.addTailer(containerID, tailer)
 }
 
 // stopTailer stops the tailer matching the containerID.
 func (l *Launcher) stopTailer(containerID string) {
 	if tailer, isTailed := l.tailers[containerID]; isTailed {
 		go tailer.Stop()
-		delete(l.tailers, containerID)
+		l.removeTailer(containerID)
+	}
+}
+
+func (l *Launcher) restartTailer(containerID string) {
+	log.Warnf("Tailing will restart for %v", ShortContainerID(containerID))
+	backoffDuration := backoffInitialDuration
+	cumulatedBackoffDuration := 0 * time.Second
+	var tailer *Tailer
+	var source *config.LogSource
+
+	for {
+		if backoffDuration > backoffMaxDuration {
+			log.Warnf("Could not restart container %v. Stop tailing", ShortContainerID(containerID))
+			return
+		}
+
+		oldTailer, exists := l.tailers[containerID]
+		if exists {
+			source = oldTailer.source
+			oldTailer.Stop()
+			l.removeTailer(containerID)
+		}
+
+		tailer = NewTailer(l.cli, containerID, source, l.pipelineProvider.NextPipelineChan(), l.erroredContainerID)
+
+		// compute the offset to prevent from missing or duplicating logs
+		since, err := Since(l.registry, tailer.Identifier(), service.Before)
+		if err != nil {
+			log.Warnf("Could not recover tailing from last committed offset, start tailing from now for container %v: %v", ShortContainerID(containerID), err)
```

Comment:
This could probably be simplified a bit too avoid being too verbose, maybe something along the lines of "Could not recover last committed offset for container %v" ?